doc: move to Sphinx

ru-fu opened the pull request:
Adding the files necessary to start building the docs in Sphinx. The output is currently available at https://lxd-sphinx.readthedocs.io/en/move-to-sphinx/

The docs should still build fine in the current format. Some toctrees will be visible as code blocks in the output.

The only content changes are on the index page. The current structure does not work in Sphinx, because you cannot have a start page that has both subpages and siblings. Therefore, I moved the start page out of Getting started and moved the installation instructions to a separate topic under Getting started. Also, there was one FAQ on the start page and one under Getting started, so I combined those.

Closes #9566

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

ru-fu commented on doc/doc-cheat-sheet.md:
```diff
@@ -0,0 +1,545 @@
+---
```
This file is currently not included in the navigation, but it can be accessed directly by URL.

ru-fu commented on doc/external_resources.md:
```diff
@@ -0,0 +1,8 @@
+# External resources
```
Files like these are required to create the navigation in Sphinx. They aren't included in the current format.

ru-fu commented on doc/index.md:
```diff
@@ -23,117 +25,6 @@ Static analysis     | GoReport              | [![Go Report Card](https://gorepor
 Translations        | Weblate               | [![Translation status](https://hosted.weblate.org/widgets/linux-containers/-/svg-badge.svg)](https://hosted.weblate.org/projects/linux-containers/lxd/)
 Project status      | CII Best Practices    | [![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/1086/badge)](https://bestpractices.coreinfrastructure.org/projects/1086)
 
-## Installing LXD from packages
```
Moved this content to a separate page - will need to add a redirect (how is this currently done?).

stgraber replied:
One thing to be careful about when messing with `index.md` is that this is also the README showed at https://github.com/lxc/lxd so we need to make sure that this one renders perfectly on Github and that it still has everything that someone needs to understand what LXD is at a glance and how to make use of the repo they're looking at.

If this becomes too problematic, we could revert to having a completely separate README.md for that purpose.

ru-fu replied:
Which is a good argument for keeping this file small and concise. Right now, it's a collection of information that doesn't fit anywhere else.
Imo, this file should contain the logo, the general intro to what LXD is, project information (status, bug reporting/contributing/support) and links to where to find more information. Installation should be moved out to a separate page (and linked, of course). And having an FAQ in two different places is extremely confusing.

But sure, we can take that in a separate PR.

ru-fu commented on doc/index.md:
```diff
@@ -149,9 +40,6 @@ Canonical RBAC.
 
 More details are [available here](security.md).
 
-## Getting started with LXD
```
Duplication - not needed (and I doubt anybody has a bookmark to this section, so no redirect needed).

ru-fu commented on doc/index.md:
```diff
@@ -169,135 +57,6 @@ find and subscribe to those at: <https://lists.linuxcontainers.org>
 ### IRC
 If you prefer live discussions, you can find us in [#lxc](https://kiwiirc.com/client/irc.libera.chat/#lxc) on irc.libera.chat.
 
-## FAQ
```
Moved this content to the existing FAQ page - will need to add a redirect (how is this currently done?).

ru-fu commented on doc/index.md:
````diff
@@ -358,3 +117,19 @@ The `hello-ubuntu.json` file referenced above could contain something like:
+
+
+
+```{toctree}
````
This toctree will be visible in the current documentation (but it is at the bottom, so probably not a big deal).

ru-fu commented on doc/metadata.yaml:
```diff
@@ -1,12 +1,16 @@
 site_logo_url: https://linuxcontainers.org/static/img/containers.png
 navigation:
-    - title: Getting started
+    - title: LXD
```
New entry for the start page.

ru-fu commented on doc/metadata.yaml:
```diff
@@ -1,12 +1,16 @@
-    - title: Getting started
+    - title: LXD
       location: index.md
+    - title: Getting started
```
Removed the link on the "Getting started" group, to be consistent with other groups. The link will still work (since it was index.md).

ru-fu commented on doc/metadata.yaml:
```diff
@@ -17,7 +21,6 @@ navigation:
           location: contributing.md
 
     - title: Configuration
-      location: configuration.md
```
Removed the link on the Configuration group to be consistent with other grouups. The content was outdated anyway and did not add anything.
In my local build, the page is still created, so if it works the same way in the real build links will not break.

stgraber commented:
Can you open another PR with just:
 - https://github.com/lxc/lxd/pull/9556/commits/5b3bc475b536c197de81a6661b3c0764e8c7d3b2
 - https://github.com/lxc/lxd/pull/9556/commits/5eddc2853b37ca46cd3ccfe26aff2f9b969d3f48

Those two look safe to merge on their own already.

For https://github.com/lxc/lxd/pull/9556/commits/099aa5e15d7560a876039eadc59355e906a5ef9e, I don't think we need all that stuff inside our doc/ directory as that's causing quite a lot of code duplication (with the theme) and is likely to mostly look confusing. I think I'd prefer we put a Github Action in place (.github/workflows/sphinx.yml) which will:
 - Setup a venv
 - Download the dependencies
 - Clone the vanilla theme
 - Build the sphinx docs
 - Pack the result as an artifact

This should avoid having to ship a lot of those files directly in the repo and should make maintaining things a bit easier.

ru-fu commented:
> Can you open another PR with just:

https://github.com/lxc/lxd/pull/9562

> This should avoid having to ship a lot of those files directly in the repo and should make maintaining things a bit easier.

I agree it would be good to leave out the Vanilla files, but most of the other files need some tweaks and are therefore required.
For the Vanilla files, how will that work for local builds? Do we need to duplicate the GitHub Action in the Makefile? Or should we add scripts that work both for building locally and through the action?

ru-fu commented:
I updated the Makefile now to download the theme, so this can be done locally and through GitHub actions.
I also added a workflow - not quite happy with that yet (I would want it to fail when there are warnings, but still upload the doc output), but it's a first attempt.

ru-fu commented:
> I also added a workflow - not quite happy with that yet (I would want it to fail when there are warnings, but still upload the doc output), but it's a first attempt.

Fixed now. I couldn't figure out a way though to output the warnings.txt file in an easily accessible location.

stgraber commented:
Moved back to Draft because of the conflicts and most of the immediately useful bits having been split up.

stgraber commented:
For the swagger API, I'd prefer it if we could git clone the repo that has the original JavaScript and CSS and use that in our build rather than embedding our own copies in the repo.

ru-fu commented:
> For the swagger API, I'd prefer it if we could git clone the repo that has the original JavaScript and CSS and use that in our build rather than embedding our own copies in the repo.

Makes sense - but we'll probably need to tweak the CSS.

stgraber commented:
> > For the swagger API, I'd prefer it if we could git clone the repo that has the original JavaScript and CSS and use that in our build rather than embedding our own copies in the repo.
> 
> Makes sense - but we'll probably need to tweak the CSS.

Hopefully we can have it include a secondary CSS or something so we don't need to actually alter the files coming from upstream.

ru-fu commented:
All changes merged through other PRs.
